lib/rbac.ex

```diff
@@ -37,4 +37,117 @@ defmodule RBAC do
   def transform_role_list_to_string(roles) do
     [Map.delete(roles, :__meta__)] |> transform_role_list_to_string()
   end
+
+  @doc """
+  `get_approles/2` fetches the roles for the app
+  """
+  def get_approles(auth_url, client_id) do
+    url = "#{auth_url}/approles/#{client_id}"
+    HTTPoison.start()
+
+    HTTPoison.get(url)
+    |> parse_body_response()
+  end
+
+  @doc """
+  `parse_body_response/1` parses the response
+  so your app can use the resulting JSON (list of roles).
+  """
+  @spec parse_body_response({atom, String.t()}) :: String.t()
+  def parse_body_response({:error, err}), do: {:error, err}
+
+  def parse_body_response({:ok, response}) do
+    body = Map.get(response, :body)
+    # IO.inspect(body)
+    # make keys of map atoms for easier access in templates
+    if body == nil do
+      {:error, :no_body}
+    else
+      {:ok, str_key_map} = Jason.decode(body)
+
+      atom_key_map =
+        Enum.map(str_key_map, fn role ->
+          for {key, val} <- role, into: %{}, do: {String.to_atom(key), val}
+        end)
+
+      {:ok, atom_key_map}
+    end
+
+    # https://stackoverflow.com/questions/31990134
+  end
+
+  @doc """
+  `init_roles/2 fetches the list of roles for an app 
+  from the auth app (auth_url) based on the client_id
+  and caches the list for fast access. 
+  ETS is an in-memory cache you get for *Free* in Elixir/Erlang.
+  See: https://elixir-lang.org/getting-started/mix-otp/ets.html
+  and: https://elixirschool.com/en/lessons/specifics/ets
+  """
+  def init_roles_cache(auth_url, client_id) do
+    {:ok, roles} = RBAC.get_approles(auth_url, client_id)
+    :ets.new(:roles_cache, [:set, :protected, :named_table])
+    # insert full list:
+    :ets.insert(:roles_cache, {"roles", roles})
+    # insert individual roles for fast lookup:
+    Enum.each(roles, fn role ->
+      :ets.insert(:roles_cache, {role.name, role})
+      :ets.insert(:roles_cache, {role.id, role})
+    end)
+  end
+
+  @doc """
+  `get_role_from_cache/1 retrieves a role from ets cache
+  """
+  def get_role_from_cache(term) do
+    case :ets.lookup(:roles_cache, term) do
+      # not found:
+      [] -> %{id: 0}
+      # extract role:
+      [{_term, role}] -> role
+    end
+  end
+
+  @doc """
+  `has_role?/2 confirms if the person has the given role
+  e.g: 
+  has_role?(conn, "home_admin") > true
+  has_role?(conn, "potus") > false
+  """
+  def has_role?(conn, role_name) do
+    role = get_role_from_cache(role_name)
+
+    person_roles =
+      conn.assigns.person.roles
```

Comment:
I personally don't like passing `conn` here.

+ It restricts you to only checking for the currently logged in user.
+ It restricts you to using `auth_plug`
+ You can't check roles of any user you want
+ You have to use this with Phoenix/Plug frameworks

Reply:
@th0mas yeah, this is a shorthand for our _current_ use-case. 
totally agree that it's not as versatile (_or as simple_) as passing the two pieces of data to be checked.
Please feel free to refactor to accept just the roles (`String` of comma-separated `ints` e.g: `"1,7,42"`) and role `"admin"`

I just thought for our _current_ use-case it would be less verbose than having to type the following:
```elixir
RBAC.has_role?(conn.assigns.person.roles, "admin") 
```
each time ... 💭 

I feel that the _context_ where this will be used is 99% "controllers" in Phoenix.
Unless you think we can use it in the Smart Home Firmware? 🤔

Reply:
It will mostly be used in Phoenix, when identifying users through an NFC tag we look them up in the database separately, which means using `conn` won't work there. It also won't work when using phoenix channels.

Reply:
Ok. cool. we need to keep our Channels options open so we may want to have a pattern match in the function so that it can accept the `conn` struct as well as a string. 💭

Reply:
Best of both worlds:
https://github.com/dwyl/rbac/blob/1ea682cc39be950a0c54ed6f2eef3d59da603807/lib/rbac.ex#L134-L142